File: Python/graph_algorithms/06_shortest_path.py
import sys
from collections import defaultdict, deque


class Graph(object):
    def __init__(self, edge_list, nr_nodes):
        self._adjacency_list = self._convert_edgelist_to_adjacency_list(edge_list)
        self._nr_nodes = nr_nodes

    def _convert_edgelist_to_adjacency_list(self, edge_list):
        adjacency_list = defaultdict(list)
        for node_from, node_to in edge_list:
            adjacency_list[node_from].append(node_to)
            adjacency_list[node_to].append(node_from)
        return dict(adjacency_list)

    def breadth_first_search(self, start_node):
        distances = [-1] * self._nr_nodes
        distances[start_node-1] = 0
        queue = deque([start_node])
        while queue:
            node = queue.popleft()
            neighbours = self._adjacency_list.get(node, [])
            for neighbour in neighbours:
                if distances[neighbour-1] == -1:
                    queue.append(neighbour)
                    distances[neighbour-1] = distances[node-1] + 1
        return distances


def find_shortest_path(edge_list, nr_nodes, s, t):
    graph = Graph(edge_list, nr_nodes)
    distances_from_s = graph.breadth_first_search(s)
    distance_s_t = distances_from_s[t-1]
    return distance_s_t


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def read_edges(reader, nr_edges):
    edges = []
    for _ in range(nr_edges):
        edge = tuple(convert_to_intlist(next(reader)))
        edges.append(edge)
    return edges


def main():
    reader = sys.stdin
    nr_nodes, nr_edges = convert_to_intlist(next(reader))
    edge_list = read_edges(reader, nr_edges)
    s, t = convert_to_intlist(next(reader))
    result = find_shortest_path(edge_list, nr_nodes, s, t)
    print(result)


if __name__ == "__main__":
    main()

File: Rust/graph_algorithms/src/bin/06_shortest_path.rs
use std::collections::{HashMap, HashSet, VecDeque};
use std::io;

struct Edge {
    start: i32,
    end: i32,
}

struct Graph {
    nr_nodes: i32,
    adjacency_list: HashMap<i32, HashSet<i32>>
}

impl Graph {
    fn from_edge_list(nr_nodes: i32, edges: &[Edge]) -> Self {
        let mut adjacency_list = HashMap::new();
        for &Edge { start, end } in edges {
            adjacency_list.entry(start).or_insert_with(HashSet::new).insert(end);
            adjacency_list.entry(end).or_insert_with(HashSet::new).insert(start);
        }

        Graph { nr_nodes, adjacency_list }
    }
}

fn convert_to_int_vector(line: &str) -> Vec<i32> {
    line
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect()
}

fn read_pair() -> (i32, i32) {
    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();

    let xs: Vec<i32> = convert_to_int_vector(&line);
    let [a, b] = xs.as_slice() else {
        panic!("Malformed input: {}", line);
    };
    (*a, *b)
}

fn read_edges(nr_edges: usize) -> Vec<Edge> {
    let mut edges = Vec::with_capacity(nr_edges);
    for _ in 0..nr_edges {
        let (start, end) = read_pair();
        edges.push(Edge { start, end });
    }

    edges
}

fn find_shortest_paths(graph: Graph, start_node: i32, target_node: i32) -> i32 {
    let mut distances = vec![-1; graph.nr_nodes as usize];
    distances[start_node as usize - 1] = 0;
    let mut queue = VecDeque::new();
    queue.push_back(start_node);
    while let Some(node) = queue.pop_front() {
        if let Some(neighbors) = graph.adjacency_list.get(&node) {
            for &neighbor in neighbors {
                if distances[neighbor as usize - 1] == -1 {
                    queue.push_back(neighbor);
                    distances[neighbor as usize - 1] = distances[node as usize - 1] + 1;
                }
            }
        }
    }

    distances[target_node as usize - 1]
}

fn main() {
    let (nr_nodes, nr_edges) = read_pair();
    let edge_list = read_edges(nr_edges as usize);
    let (u, v) = read_pair();
    let graph = Graph::from_edge_list(nr_nodes, &edge_list);
    let result = find_shortest_paths(graph, u, v);
    println!("{}", result);
}
